src/mon/PGMap.cc

```diff
@@ -1212,10 +1212,12 @@ void PGMap::apply_incremental(CephContext *cct, const Incremental& inc)
       stat_osd_sub(t->first, t->second);
       osd_stat.erase(t);
     }
-    for (auto i = pool_statfs.begin();  i != pool_statfs.end(); ++i) {
+    for (auto i = pool_statfs.begin();  i != pool_statfs.end();) {
       if (i->first.second == *p) {
 	pg_pool_sum[i->first.first].sub(i->second);
-	pool_statfs.erase(i);
+	i = pool_statfs.erase(i);
```

Comment:
From core PR scrub: this looks like a valid issue. We have a tracker for that?

Reply:
@rzarzynski Yes, they've been rather unnoticed so far. https://tracker.ceph.com/issues/58215 & https://tracker.ceph.com/issues/58303

Reply:
Thanks, @osnyx! I linked these tickets.

Would you mind to rework the commit description by:
1) putting the `Fixes: https://tracker.ceph.com/issues/58303` line just before the `Signed-off-by`,
2) changing the title to: `mon: fix iterator mishandling in PGMap::apply_incremental`?

Reply:
Let's take a look on the `pool_statfs` member:

```cpp
  typedef mempool::pgmap::map<
    std::pair<int64_t, int>,  // <pool, osd>
    store_statfs_t>
      per_osd_pool_statfs_t;

  per_osd_pool_statfs_t pool_statfs;
```

The underlying type is `std::map`. From [the doc](https://en.cppreference.com/w/cpp/container/map/erase):

> References and iterators to the erased elements are invalidated. Other references and iterators are not affected. 

My understanding is that we shall not `i++` if the `i`-pointed element has been erased. This stays coherent with the API design (returning new iterator) and even with the example there:

```cpp
    // erase all odd numbers from c
    for (auto it = c.begin(); it != c.end();)
    {
        if (it->first % 2 != 0)
            it = c.erase(it);
        else
            ++it;
    }
```